internal/server/tls.go

```diff
@@ -17,20 +16,50 @@ import (
 	"github.com/infrahq/infra/internal/logging"
 )
 
+// MapCache is a simple in-memory caching mechanism, it is not thread safe
+type MapCache map[string][]byte
+
+func (m MapCache) Get(ctx context.Context, name string) ([]byte, error) {
+	data, cached := m[name]
+	if !cached {
+		return nil, autocert.ErrCacheMiss
+	}
+	return data, nil
+}
+
+func (m MapCache) Put(ctx context.Context, name string, data []byte) error {
+	m[name] = data
+	return nil
+}
+
+func (m MapCache) Delete(ctx context.Context, name string) error {
+	delete(m, name)
+	return nil
+}
+
+func (m MapCache) getKeyPair(ctx context.Context, serverName string) (cert, key []byte) {
+	certBytes, _ := m.Get(ctx, serverName+".crt")
+	keyBytes, _ := m.Get(ctx, serverName+".key")
+	if certBytes == nil || keyBytes == nil {
+		logging.Infof("no cached TLS cert for %v", serverName)
+	}
+	return certBytes, keyBytes
+}
+
+var certCache = make(MapCache)
+
+var ca keyPair
+
+var cacheLock sync.RWMutex
```

Comment:
Same with these. I think they are better off as local state in a closure. We could also create a struct if you prefer that over a closure.

Reply:
I brought back the closure 👍